internal/observers/ecs/ecs.go

```diff
@@ -0,0 +1,293 @@
+//  - type: ecs
+// monitors:
+//  - type: prometheus-exporter
+//    discoveryRule: name == "app" && port == 8080
+//    intervalSeconds: 1
+// ```
+//
+// And given docker labels as follows (remember that discovery rules are
+// irrelevant to endpoints that specify `monitorType` labels):
+//
+//  - `agent.signalfx.com.config.8080-app.metricPath`: `/appMetrics`
+//  - `agent.signalfx.com.config.8080-goruntime.metricPath`: `/goMetrics`
+//
+// Would result in the `app` endpoint getting an interval of 1 second and the
+// `goruntime` endpoint getting the default interval of the agent.
+
+// ENDPOINT_TYPE(ContainerEndpoint): true
+
+var logger = log.WithFields(log.Fields{"observerType": observerType})
+
+// Config for the ecs observer
+type Config struct {
+	config.ObserverConfig
+	// The URL of the ECS task metadata. Default is http://169.254.170.2/metadata, which is hardcoded by AWS for version 2.
+	MetadataEndpoint string `yaml:"metadataEndpoint" default:"http://169.254.170.2/metadata"`
+	// A mapping of container label names to dimension names that will get
+	// applied to the metrics of all discovered services. The corresponding
+	// label values will become the dimension values for the mapped name.  E.g.
+	// `io.kubernetes.container.name: container_spec_name` would result in a
+	// dimension called `container_spec_name` that has the value of the
+	// `io.kubernetes.container.name` container label.
+	LabelsToDimensions map[string]string `yaml:"labelsToDimensions"`
+}
+
+// ECS observer plugin
+type ECS struct {
+	serviceCallbacks *observers.ServiceCallbacks
+	serviceDiffer    *observers.ServiceDiffer
+	config           *Config
+}
+
+func init() {
+	observers.Register(observerType, func(cbs *observers.ServiceCallbacks) interface{} {
+		return &ECS{
+			serviceCallbacks: cbs,
+		}
+	}, &Config{})
+}
+
+// Configure the ecs observer
+func (o *ECS) Configure(config *Config) error {
+	o.config = config
+
+	if o.serviceDiffer != nil {
+		o.serviceDiffer.Stop()
+	}
+
+	o.serviceDiffer = &observers.ServiceDiffer{
+		DiscoveryFn:     o.discover,
+		IntervalSeconds: 5,
+		Callbacks:       o.serviceCallbacks,
+	}
+	o.serviceDiffer.Start()
+
+	return nil
+}
+
+// Discover services from ECS task metadata endpoint
+func (o *ECS) discover() []services.Endpoint {
+	client := &http.Client{
+		Timeout: 10 * time.Second,
+	}
+
+	response, err := client.Get(o.config.MetadataEndpoint)
+	if err != nil {
+		logger.WithError(err).Error("Could not connect to metadata endpoint")
+		return nil
+	}
+	defer response.Body.Close()
+
+	if response.StatusCode != http.StatusOK {
+		logger.WithError(err).Error("Could not receive ECS metadata : %s ", http.StatusText(response.StatusCode))
+		return nil
+	}
+
+	body, err := ioutil.ReadAll(response.Body)
+	if err != nil {
+		logger.WithError(err).Error("failed to read ECS metadata response body - %v", err)
+		return nil
+	}
+
+	var metadata ecs.TaskMetadata
+
+	if err := json.Unmarshal(body, &metadata); err != nil {
```

Comment:
You can do the decoding like this and it's a bit simpler:

```
if err := json.NewDecoder(response.Body).Decode(&metadata); err != nil {
```

then you don't need the intermediate `body` var or error check.